subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/InputProperties.java

```diff
@@ -66,6 +67,15 @@ private static boolean isBinaryComparableProperty(Object inputProperty) {
             Object item = processingQueue.pop();
 
             Class cls = item.getClass();
+            try {
+                // Given the equals wasn't override, use binary comparison.
+                if (cls.getMethod("equals", Object.class).getDeclaringClass().equals(Object.class)) {
+                    continue;
```

Comment:
I guess we could also have logic in place that asks if the method even exists before invoking it.

    cls.getMethod("equals", new Class[] { Object.class });

That would make the maybe a bit more readable. I think the `continue` is kind of weird.